src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -390,8 +398,7 @@ private AsyncClusterWriter buildWriter(final File outputPrefix) {
             molecularBarcodeFiles[i] = new File(outputDir, String.format("%s.index_%d.%s", prefixString, i + 1, suffixString));
         }
 
-
-        return new AsyncClusterWriter(new ClusterToFastqWriter(templateFiles, sampleBarcodeFiles, molecularBarcodeFiles), 1024);
+        return writerPool.pool(new ClusterToFastqWriter(templateFiles, sampleBarcodeFiles, molecularBarcodeFiles), new ArrayBlockingQueue<>(20000), 20000) ;
```

Comment:
I think I'd suggest doing setting this based on `MAX_RECORDS_IN_RAM` by:
1. Changing `populateWritersFromMultiplexParams` to load up all the rows first so it knows how many samples we have
2. Setting the queue bounds as `MAX_RECORDS_IN_RAM / num_samples`
3. Setting the write threshold to be `0.5 * queue_size` as a starting point.

Reply:
`LinkedBlockingQueue` was preferable because each end of the queue had seperate locks. So you could add to one end while pulling off the other without contention. I'm not sure if `ArrayBlockingQueue` can do that. 

I did also prefer unbounded, but if `LinkedBlockingQueue` can be bounded it's worth a try. Agreed on making the write threshold less the queue limit to avoid waiting on it. Maybe 2x the queue limit over the write limit?

In an ideal world the unbounded queue would work for you, but I'm sure that's only true up to some magic ratio of threads : writers, which we are apparently well over.